restore tipannotations; fix #53

BenjaminDoran opened the pull request:
This should fix issue #53,  

### what I had to do
There may be a better way. but I put in a new `@series begin ... end` block with changes below, which seems to show the annotations again.

It required adding Plots as a dependency to get access to the Plots.text function. Pending issue https://github.com/JuliaPlots/RecipesBase.jl/issues/72 this seemed to be the only way to do it.

#### dendrogram example

```julia
if dend.showtips
    @series begin
        seriestype := :scatter
        markersize := 0
        series_annotations := map(x -> text(x[3], :left, dend.tipfont...), dend.tipannotations)
        getindex.(dend.tipannotations, 1), getindex.(dend.tipannotations, 2)
    end
end
```

#### fan example

```julia
if fan.showtips
    @series begin
        seriestype := :scatter
        markersize := 0
        series_annotations := map(x -> text(x[3], :left, rad2deg(adjust(x[2])), fan.tipfont...), fan.tipannotations)
        xys = map(x-> (_tocirc.(x[1], adjust.(x[2]))), fan.tipannotations)
        first.(xys), last.(xys)
    end
end
```

### Minimum example
```julia
using Phylo
using StatsPlots
tree = rand(Ultrametric(10))
plot(tree, treetype=:dendrogram, size=(400,400))
tree = rand(Ultrametric(10))
plot(tree, treetype=:fan, size=(400,400))
```

![image](https://user-images.githubusercontent.com/20494558/124536693-04ec3800-ddde-11eb-999a-5f8bafbdbfad.png)

![image](https://user-images.githubusercontent.com/20494558/124536667-f867df80-dddd-11eb-8d5c-2ef014bcd169.png)

mkborregaard commented:
Great, I'll have a look. I can say immediately that adding a Plots dep is not acceptable, but I don't think that should be necessary either.

mkborregaard commented on src/plot.jl:
```diff
@@ -85,7 +86,14 @@ struct Fan; x; y; tipannotations; marker_x; marker_y; showtips; tipfont; marker_
+        @series begin
+            seriestype := :scatter
+            markersize := 0
+            series_annotations := map(x -> text(x[3], :left, dend.tipfont...), dend.tipannotations)
```
do you need the `text` here?

BenjaminDoran replied:
for this solution, yes. It writes out the unformatted entire tuple otherwise.

mkborregaard commented:
I can't work out why the original code doesn't work as it is. Consider
```julia
struct MyType end

@recipe function f(x::MyType)
    annotations := [
        (2, 4, ( "test", :right, 20)), 
        (3, 4, ("test2", :left, 10))
    ]
    1:10, 1:10
end

plot(MyType())
```
![Skærmbillede 2021-07-06 kl  15 43 35](https://user-images.githubusercontent.com/8429802/124610602-f26f0e80-de70-11eb-9163-efe5ce255d1a.png)

BenjaminDoran commented:
Thanks for the test example, I don't really know why, but changing the location of the `annotations` argument fixes the issue.

i.e. 

#### Doesn't work
```julia
struct MyType end
struct myseries; x; y; end

@recipe function f(x::MyType)
    seriestype := :myseries
    myseries(1:10, 1:10)
end


@recipe function f(ms::myseries)
    @series begin
        seriestype := :path
        ms.x, ms.y
    end
    annotations := [(2, 4, ( "test", :right, 20)), (3, 4, ("test2", :left, 10))]
    nothing
end

plot(MyType())
```
#### works
```julia
struct MyType end
struct myseries; x; y; end

@recipe function f(x::MyType)
    seriestype := :myseries
    myseries(1:10, 1:10)
end


@recipe function f(ms::myseries)
    annotations := [(2, 4, ( "test", :right, 20)), (3, 4, ("test2", :left, 10))]
    @series begin
        seriestype := :path
        ms.x, ms.y
    end
    nothing
end

plot(MyType())
```
I agree it is a much better solution, than needing the plots dependency. I reverted my commit and just moved your code blocks to above the other series blocks. And, it seems to work now!

![image](https://user-images.githubusercontent.com/20494558/124645771-73baa700-de59-11eb-84a8-3bb67b68385c.png)

BenjaminDoran commented on src/plot.jl:
```diff
@@ -86,14 +88,6 @@ struct Fan; x; y; tipannotations; marker_x; marker_y; showtips; tipfont; marker_
-        end
-    end
     primary = false
     label = ""
```
also what are these two lines doing? are they meant to be ``primary := false; label := ""``? or just deleted?

mkborregaard replied:
`primary := false; label := ""` as you say :+1:

mkborregaard replied:
do you mind updating that before I merge?

mkborregaard commented:
Awesome!

mkborregaard commented:
Thanks a lot @BenjaminDoran !

richardreeve commented:
Yes, thanks so much for fixing this - knowing nothing about plotting it has just been sitting around mildly annoying me for ages! Looking at the PR, it's not remotely clear to me why this solution works either, but never mind :)
